lxd/storage/drivers/driver_pure_util.go

```diff
@@ -517,6 +572,475 @@ func (p *pureClient) disconnectHostFromVolume(poolName string, volName string, h
+	if err != nil {
+		return err
+	}
+
+	unlock, err := locking.Lock(d.state.ShutdownCtx, d.config["pure.mode"])
+	if err != nil {
+		return err
+	}
+
+	defer unlock()
+
+	switch d.config["pure.mode"] {
+	case pureModeISCSI:
+		iqn, err = d.hostIQN()
+		if err != nil {
+			return err
+		}
+
+		host, err = d.client().getHostByIQN(iqn)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Disconnect the volume from the host and ignore error if connection does not exist.
+	err = d.client().disconnectHostFromVolume(vol.pool, volName, host.Name)
+	if err != nil && !api.StatusErrorCheck(err, http.StatusNotFound) {
+		return err
+	}
+
+	// Wait until the volume has disappeared.
+	volumePath, _, _ := d.getMappedDevPath(vol, false)
+	if volumePath != "" {
+		if d.config["pure.mode"] == pureModeISCSI {
+			// When volume is disconnected from the host, the device will remain on the system.
+			//
+			// To remove the device, we need to either logout from the session or remove the
+			// device manually. Logging out of the session is not desired as it would disconnect
+			// from all connected volumes. Therefore, we need to manually remove the device.
+			split := strings.Split(filepath.Base(volumePath), "/")
+			devName := split[len(split)-1]
+
+			path := fmt.Sprintf("/sys/block/%s/device/delete", devName)
+			if shared.PathExists(path) {
+				err := os.WriteFile(path, []byte("1"), 0400)
+				if err != nil {
+					return fmt.Errorf("Failed to unmap volume %q: Failed to remove device %q: %w", vol.name, devName, err)
+				}
+			}
+		}
+
+		ctx, cancel := context.WithTimeout(d.state.ShutdownCtx, 10*time.Second)
+		defer cancel()
+
+		if !waitGone(ctx, volumePath) {
+			return fmt.Errorf("Timeout exceeded waiting for PureStorage volume %q to disappear on path %q", vol.name, volumePath)
+		}
+	}
+
+	// If this was the last volume being unmapped from this system, terminate iSCSI session
+	// and remove the host from PureStorage.
+	if host.ConnectionCount == 1 {
+		target, err := d.client().getTarget()
+		if err != nil {
+			return err
+		}
+
+		// Disconnect from the target.
+		err = d.disconnect(*target)
+		if err != nil {
+			return err
+		}
+
+		// Remove the host from PureStorage.
+		err = d.client().deleteHost(host.Name)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getMappedDevPath returns the local device path for the given volume.
+// Indicate with mapVolume if the volume should get mapped to the system if it isn't present.
+func (d *pure) getMappedDevPath(vol Volume, mapVolume bool) (string, revert.Hook, error) {
+	revert := revert.New()
+	defer revert.Fail()
+
+	if mapVolume {
+		err := d.mapVolume(vol)
+		if err != nil {
+			return "", nil, err
+		}
+
+		revert.Add(func() { _ = d.unmapVolume(vol) })
+	}
+
+	// findDevPathFunc has to be called in a loop with a set timeout to ensure
+	// all the necessary directories and devices can be discovered.
+	findDevPathFunc := func(diskPrefix string, diskSuffix string) (string, error) {
```

Comment:
or can at least parts of it be shared with powerflex?

Reply:
Yes this part can be shared with PowerFlex, however, we use different prefix/suffix to locate a volume.

Reply:
Would it be acceptable if I address shared functionality with PowerFlex in a separate smaller PR afterwards? My point being is that messing with PowerFlex without being able to test seems a bit risky, but also do not want to put burden on Julian to keep testing PowerFlex after each change to this big PR.